### PR TITLE
Add price signals and cross-symbol signal support

### PR DESCRIPTION
**Pull request by michaelchu**
## Summary

- Add 14 new first-class price signals in `optopsy/signals/price.py`: price level (above/below/cross), gaps, N-period breakouts, daily returns, drawdown/rally from high/low, and consecutive up/down streaks
- Add cross-symbol signal support via `signal_symbol` parameter on `build_signal` and `build_custom_signal`, enabling use cases like "use VIX data to trigger entries on SPX options"
- Update agent system prompt, tool schemas, and model descriptions to document all new signals and the cross-symbol workflow

## Test plan

- [x] 28 new tests in `tests/test_signals_price.py` covering all 14 signals + edge cases
- [x] Full test suite passes (1862 tests)
- [x] `ruff check` and `ruff format` clean
- [x] `ty` type checker passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: optopsy/ui/tools/_signals_builder.py [hunk 1]
```diff
@@ -83,13 +84,18 @@ def _handle_build_signal(arguments, dataset, signals, datasets, results, _result
...
         iv_data = _iv_signal_data(dataset)
         if iv_data is None:
             return _result(_IV_MISSING_MSG)
         signal_data = iv_data
     elif needs_stock:
-        signal_data = _fetch_stock_data_for_signals(dataset)
+        if signal_symbol:
+            signal_data = _fetch_stock_data_for_symbol(signal_symbol, dataset)
+        else:
+            signal_data = _fetch_stock_data_for_signals(dataset)
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_signals_builder.py:121]: Good catch. Added `signal_symbol = signal_symbol.strip()` in both `_handle_build_signal` and `_handle_build_custom_signal`. The `.upper()` normalization is already handled downstream in `_fetch_stock_data_for_symbol` (line 139). Fixed in 6057c49.

File: optopsy/ui/tools/_signals_builder.py [hunk 2]
```diff
@@ -160,6 +166,17 @@ def _handle_build_signal(arguments, dataset, signals, datasets, results, _result
...
+
+    # Cross-symbol: remap signal symbol to options dataset symbol(s)
+    if signal_symbol and not raw_signal_dates.empty:
+        options_symbols = dataset["underlying_symbol"].unique().tolist()
+        frames = []
+        for opt_sym in options_symbols:
+            mapped = raw_signal_dates.copy()
+            mapped["underlying_symbol"] = opt_sym
+            frames.append(mapped)
+        raw_signal_dates = pd.concat(frames, ignore_index=True)
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_signals_builder.py]: Valid point. The remap is now gated on `needs_stock` — it only runs when we actually fetched data for `signal_symbol`. IV signals use options data (whose `underlying_symbol` already matches the dataset), and date-only signals derive dates from the dataset itself, so remapping would incorrectly duplicate rows. Fixed in 6057c49.

File: optopsy/ui/tools/_signals_builder.py [hunk 3]
```diff
@@ -160,6 +166,17 @@ def _handle_build_signal(arguments, dataset, signals, datasets, results, _result
...
+
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_signals_builder.py:210]: Acknowledged. The cross-symbol remapping logic itself is straightforward (copy + replace `underlying_symbol`), and the signal-level functions are covered in `test_signals_price.py`. Adding integration-level tests for the full build_signal → remap → intersect workflow would require mocking yfinance + the cache layer. We'll track this as a follow-up for broader UI tool test coverage.

File: optopsy/ui/tools/_helpers.py [hunk 4]
```diff
@@ -118,6 +118,58 @@ def _fetch_stock_data_for_signals(dataset: pd.DataFrame) -> pd.DataFrame | None:
...
+                "open",
+                "high",
+                "low",
+                "close",
+                "volume",
+            ]
+        ]
+    except (OSError, ValueError, KeyError, pd.errors.ParserError) as exc:
+        _log.warning("yfinance fetch failed for %s: %s", symbol, exc)
+        return None
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_helpers.py]: Acknowledged. `_fetch_stock_data_for_symbol` follows the same pattern as `_fetch_stock_data_for_signals` (which also lacks direct unit tests due to the yfinance dependency). The column contract and date slicing logic are identical. We'll track adding mocked integration tests for both fetch helpers as a follow-up.

File: optopsy/ui/tools/_schemas.py [hunk 5]
```diff
@@ -616,6 +635,10 @@ def _normalize_days_param(days):
         "psar_sell",
         "chop_above",
         "chop_below",
+        "gap_up",
+        "gap_down",
+        "high_of_n_days",
+        "low_of_n_days",
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_schemas.py]: Good catch. Added `_OPEN_SIGNALS = frozenset({"gap_up", "gap_down"})` in `_schemas.py` and a validation block in `_handle_build_signal` that checks for the `open` column when these signals are present, returning a clear error message. Fixed in 24f9129.

File: optopsy/ui/tools/_helpers.py
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_helpers.py:174]: Already addressed in d909f81 — extracted `_fetch_single_symbol_stock(symbol, padded_start, date_max)` as a shared helper. Both `_fetch_stock_data_for_signals` and `_fetch_stock_data_for_symbol` now delegate to it, eliminating the duplicated cache-read/fetch/slice/column-select logic.

File: tests/test_signals_price.py [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+        """gap_up should fire when open > prev close * (1 + pct/100)."""
+        # prev close: [-, 102, 103, 101, 104, 98, 97, 101]
+        # open:       [100, 102, 101, 105, 99, 98, 100, 103]
+        # gap pct:    [-, 0, -1.9%, +3.96%, -4.8%, 0%, +3.1%, +1.98%]
+        # At 1% threshold: bar 3 (105 > 101*1.01=102.01) and bar 7 (103 > 101*1.01=102.01)
+        result = gap_up(pct=1.0)(ohlcv_data)
+        gap_indices = [i for i, v in enumerate(result) if v]
+        assert 3 in gap_indices  # 105 vs prev close 101
+        assert 7 in gap_indices  # 103 vs prev close 101
+
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply tests/test_signals_price.py:145]: Good point. Updated both `test_gap_up_detects_gap` and `test_gap_down_detects_gap` to assert the exact list of firing indices (`assert gap_indices == [3, 6, 7]` and `== [2, 4]`) instead of just checking membership. Fixed in 24f9129.

File: tests/test_signals_price.py [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+                "underlying_symbol": "SPX",
+                "quote_date": dates,
+                "close": [100, 101, 102, 103, 104],
+            }
+        )
+        # _get_high falls back to close, so this should still work
+        result = high_of_n_days(period=3)(data)
+        # Each bar has rising close, so every bar after warmup reaches a new high
+        assert result.any()
+
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply tests/test_signals_price.py:242]: Agreed. Replaced `assert result.any()` with `assert list(result) == [False, True, True, True, True]`, which verifies the exact bars that fire when `_get_high` falls back to close on monotonically rising data. Fixed in 24f9129.

File: optopsy/signals/price.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+    days = int(days)
+
+    def _signal(data: pd.DataFrame) -> "pd.Series[bool]":
+        def _compute_group(group: pd.DataFrame) -> "pd.Series[bool]":
+            close = _get_close(group)
+            if close is None:
+                return pd.Series(False, index=group.index)
+            up = (close > close.shift(1)).astype(int)
+            streak = up.rolling(days, min_periods=days).sum()
+            return (streak == days).fillna(False)
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/signals/price.py:270]: Added `if days < 1: raise ValueError(...)` to both `consecutive_up` and `consecutive_down`, plus a test (`test_consecutive_days_below_one_raises`) verifying the guard. Fixed in 24f9129.

File: optopsy/ui/tools/_helpers.py [hunk 9]
```diff
@@ -51,6 +51,47 @@
...
+        result = cached[
+            (pd.to_datetime(cached["date"]).dt.date >= padded_start)
+            & (pd.to_datetime(cached["date"]).dt.date <= date_max)
+        ].rename(columns={"date": "quote_date"})
+        if result.empty:
+            return None
+        return result[_STOCK_COLS]
+    except (OSError, ValueError, KeyError, pd.errors.ParserError) as exc:
+        _log.warning("yfinance fetch failed for %s: %s", symbol, exc)
+        return None
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_helpers.py:99]: Valid concern. `_fetch_single_symbol_stock` now selects only available columns (`[c for c in _STOCK_COLS if c in result.columns]`) instead of hard-coding all 7, and returns `None` only if `close` is missing. This lets signals that only need `close` work even when yfinance omits optional columns, while letting the caller-level validations (`_OHLC_SIGNALS`, `_OPEN_SIGNALS`, `_VOLUME_SIGNALS`) report specific missing-column errors. Fixed in 79c8c24.

File: optopsy/ui/tools/_schemas.py [hunk 10]
```diff
@@ -616,9 +635,16 @@ def _normalize_days_param(days):
...
     }
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_schemas.py:640]: Good point. Removed `gap_up`/`gap_down` from `_OHLC_SIGNALS` — they don't need `high`/`low` and already have their own `_OPEN_SIGNALS` validation checking for the `open` column. `high_of_n_days`/`low_of_n_days` remain in `_OHLC_SIGNALS` since they genuinely use the `high`/`low` columns. Fixed in 79c8c24.

File: tests/test_signals_price.py [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+        )
+        result = high_of_n_days(period=5)(data)
+        # Bar 7: close=103, prev 5-bar high max = max(103,99,98,100,104)
+        # Wait — rolling high is on the high column, shifted by 1
+        # Bar 7 rolling(5).max().shift(1) = max of bars 2-6 highs = max(101,103,99,98,100) = 103
+        # close=103 >= 103 → True
+        assert result.iloc[7] == True
+        # Bar 9: close=105, prev 5-bar high = max of bars 4-8 = max(99,98,100,104,103) = 104
+        # 105 >= 104 → True
+        assert result.iloc[9] == True
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply tests/test_signals_price.py]: Agreed. Updated to assert exact firing indices: `assert fire_indices == [1, 3, 7, 9]`. Added per-bar computation comments for clarity. Fixed in 79c8c24.

File: tests/test_signals_price.py [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+                "high": [100, 102, 101, 103, 99, 98, 100, 97, 96, 95],
+                "low": [98, 100, 99, 101, 97, 96, 98, 95, 94, 93],
+                "close": [99, 101, 100, 102, 98, 97, 99, 96, 95, 93],
+            }
+        )
+        result = low_of_n_days(period=5)(data)
+        # Bar 7: close=96, prev 5-bar low = min of bars 2-6 lows = min(99,101,97,96,98) = 96
+        # 96 <= 96 → True
+        assert result.iloc[7] == True
+
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply tests/test_signals_price.py:221]: Agreed. Updated to assert exact firing indices: `assert fire_indices == [4, 5, 7, 8, 9]`. Added per-bar computation comments. Fixed in 79c8c24.
